File: src/repotask/commands/query.py
"""Direct knowledge lookups: `convention`, `recipe`, `search`."""

from __future__ import annotations

from typing import Any

import typer
from rich.console import Group, RenderableType
from rich.markdown import Markdown
from rich.table import Table

from repotask.commands._context import open_knowledge
from repotask.errors import RepoTaskError
from repotask.kb.schema import Document
from repotask.kb.store import KnowledgeBase
from repotask.output import emit, guard


def _document_data(document: Document) -> dict[str, Any]:
    return {
        "id": document.id,
        "title": document.title,
        "layer": document.layer,
        "stacks": document.stacks,
        "tags": document.tags,
        "path": document.path,
        "body": document.body,
    }


def _render_documents(data: dict[str, Any]) -> RenderableType:
    parts: list[RenderableType] = []
    for item in data["documents"]:
        parts.append(Markdown(f"# {item['title']}\n\n_{item['path']}_\n\n{item['body']}"))
    return Group(*parts) if parts else "No matching documents."


def _lookup(kb: KnowledgeBase, layer: str, topic: str, stacks: list[str]) -> list[Document]:
    """Exact id first; otherwise fall back to a stack-filtered search."""
    pool = kb.conventions if layer == "convention" else kb.recipes
    exact = pool.get(topic)
    if exact is not None:
        return [exact]
    hits = kb.search(topic, layer=layer)
    scoped = [
        hit.document
        for hit in hits
        if not hit.document.stacks or set(hit.document.stacks) & set(stacks)
    ]
    return scoped or [hit.document for hit in hits]


@guard
def convention(
    topic: str = typer.Argument(..., help="Convention id or search topic."),
    limit: int = typer.Option(3, "--limit", help="Maximum documents to return."),
) -> None:
    """Read architecture and stack conventions for this project."""
    config, kb = open_knowledge()
    documents = _lookup(kb, "convention", topic, config.project.stacks)[:limit]
    if not documents:
        raise RepoTaskError(
            f"No convention matched '{topic}'. Try `repo-task search {topic}`."
        )
    emit(
        "convention",
        {"topic": topic, "documents": [_document_data(item) for item in documents]},
        _render_documents,
    )


@guard
def recipe(
    task: str = typer.Argument(..., help="Recipe id or the task you want a playbook for."),
    limit: int = typer.Option(3, "--limit", help="Maximum documents to return."),
) -> None:
    """Read the project playbook for a specific task."""
    config, kb = open_knowledge()
    documents = _lookup(kb, "recipe", task, config.project.stacks)[:limit]
    if not documents:
        raise RepoTaskError(f"No recipe matched '{task}'. Try `repo-task search {task}`.")
    emit(
        "recipe",
        {"task": task, "documents": [_document_data(item) for item in documents]},
        _render_documents,
    )


@guard
def search(
    query: str = typer.Argument(..., help="Keywords to search across the knowledge base."),
    layer: str = typer.Option("", "--layer", help="Restrict to `convention` or `recipe`."),
    limit: int = typer.Option(20, "--limit"),
) -> None:
    """Search conventions and recipes; returns ids to read with `convention` or `recipe`."""
    if layer and layer not in {"convention", "recipe"}:
        raise RepoTaskError("--layer must be `convention` or `recipe`.")
    _config, kb = open_knowledge()
    hits = kb.search(query, layer=layer or None, limit=limit)
    data = {
        "query": query,
        "hits": [
            {
                "id": hit.document.id,
                "title": hit.document.title,
                "layer": hit.document.layer,
                "stacks": hit.document.stacks,
                "tags": hit.document.tags,
                "score": hit.score,
                "excerpt": hit.excerpt,
            }
            for hit in hits
        ],
    }
    emit("search", data, _render_hits)


def _render_hits(data: dict[str, Any]) -> RenderableType:
    if not data["hits"]:
        return f"No knowledge matched '{data['query']}'."
    table = Table(title=f"Knowledge matching '{data['query']}'")
    table.add_column("id")
    table.add_column("layer")
    table.add_column("title")
    table.add_column("excerpt", overflow="fold")
    for hit in data["hits"]:
        table.add_row(hit["id"], hit["layer"], hit["title"], hit["excerpt"])
    return table

File: src/commands/query.rs
//! Direct knowledge lookups: `convention`, `recipe`, `search`.

use anyhow::{bail, Result};
use serde_json::{json, Value};

use crate::config;
use crate::kb;
use crate::kb::schema::{Document, Layer};
use crate::kb::store::KnowledgeBase;
use crate::output;

fn document_json(document: &Document) -> Value {
    json!({
        "id": document.id,
        "title": document.title,
        "layer": document.layer.as_str(),
        "stacks": document.stacks,
        "tags": document.tags,
        "path": document.path,
        "body": document.body,
    })
}

/// Exact id first; otherwise fall back to a stack-filtered search.
fn lookup(kb: &KnowledgeBase, layer: Layer, topic: &str, stacks: &[String]) -> Vec<Document> {
    let pool = match layer {
        Layer::Convention => &kb.conventions,
        Layer::Recipe => &kb.recipes,
    };
    if let Some(exact) = pool.get(topic) {
        return vec![exact.clone()];
    }
    let hits = kb.search(topic, Some(layer), 20);
    let scoped: Vec<Document> = hits
        .iter()
        .filter(|hit| {
            hit.document.stacks.is_empty()
                || hit
                    .document
                    .stacks
                    .iter()
                    .any(|stack| stacks.contains(stack))
        })
        .map(|hit| hit.document.clone())
        .collect();
    if scoped.is_empty() {
        hits.into_iter().map(|hit| hit.document).collect()
    } else {
        scoped
    }
}

fn render_documents(value: &Value) -> String {
    let Some(documents) = value["documents"].as_array() else {
        return "No matching documents.".into();
    };
    documents
        .iter()
        .map(|document| {
            format!(
                "{}\n{}\n\n{}",
                output::style::heading(&format!(
                    "# {}",
                    document["title"].as_str().unwrap_or("")
                )),
                output::style::dim(document["path"].as_str().unwrap_or("")),
                document["body"].as_str().unwrap_or(""),
            )
        })
        .collect::<Vec<_>>()
        .join("\n\n")
}

pub fn convention(topic: &str, limit: usize) -> Result<()> {
    let config = config::load()?;
    let kb = kb::open(&config, None)?;
    let documents: Vec<Document> = lookup(&kb, Layer::Convention, topic, &config.project.stacks)
        .into_iter()
        .take(limit)
        .collect();
    if documents.is_empty() {
        bail!("No convention matched '{topic}'. Try `repo-task search {topic}`.");
    }
    let data = json!({
        "topic": topic,
        "documents": documents.iter().map(document_json).collect::<Vec<_>>(),
    });
    output::emit("convention", &data, render_documents);
    Ok(())
}

pub fn recipe(task: &str, limit: usize) -> Result<()> {
    let config = config::load()?;
    let kb = kb::open(&config, None)?;
    let documents: Vec<Document> = lookup(&kb, Layer::Recipe, task, &config.project.stacks)
        .into_iter()
        .take(limit)
        .collect();
    if documents.is_empty() {
        bail!("No recipe matched '{task}'. Try `repo-task search {task}`.");
    }
    let data = json!({
        "task": task,
        "documents": documents.iter().map(document_json).collect::<Vec<_>>(),
    });
    output::emit("recipe", &data, render_documents);
    Ok(())
}

pub fn search(query: &str, layer: &str, limit: usize) -> Result<()> {
    let layer = match layer {
        "" => None,
        "convention" => Some(Layer::Convention),
        "recipe" => Some(Layer::Recipe),
        _ => bail!("--layer must be `convention` or `recipe`."),
    };
    let config = config::load()?;
    let kb = kb::open(&config, None)?;
    let hits = kb.search(query, layer, limit);
    let data = json!({
        "query": query,
        "hits": hits.iter().map(|hit| json!({
            "id": hit.document.id,
            "title": hit.document.title,
            "layer": hit.document.layer.as_str(),
            "stacks": hit.document.stacks,
            "tags": hit.document.tags,
            "score": hit.score,
            "excerpt": hit.excerpt,
        })).collect::<Vec<_>>(),
    });
    output::emit("search", &data, |value| {
        let Some(hits) = value["hits"].as_array() else {
            return String::new();
        };
        if hits.is_empty() {
            return format!(
                "No knowledge matched '{}'.",
                value["query"].as_str().unwrap_or("")
            );
        }
        hits.iter()
            .map(|hit| {
                format!(
                    "{} {:<11} {}\n    {}",
                    output::style::id(&format!("{:<26}", hit["id"].as_str().unwrap_or(""))),
                    hit["layer"].as_str().unwrap_or(""),
                    hit["title"].as_str().unwrap_or(""),
                    output::style::dim(hit["excerpt"].as_str().unwrap_or("")),
                )
            })
            .collect::<Vec<_>>()
            .join("\n")
    });
    Ok(())
}
